File: Sorting/Bubble Sort/Lua/Bubble_Sort.lua
-- Bubble sort (also known as sinking sort) is a simple sorting algorithm that
-- repeatedly steps through the list to be sorted, compares each pair of
-- adjacent items and swaps them if they are in the wrong order. The pass
-- through the list is repeated until no swaps are needed, which indicates that
-- the list is sorted.

local function bubblesort(t)
  repeat
    local sorted = true
    for i=2, #t do
      if (t[i-1] > t[i]) then
        t[i-1], t[i] = t[i], t[i-1]
        sorted = false
      end
    end
  until sorted
end

local numbers = {5, 1, 4, 2, 8}
bubblesort(numbers)
for _, number in ipairs(numbers) do
  print(number)
end

File: Sorting/Bubble Sort/JavaScript/Bubble_Sort.js
// Function Bubble sort
// Description: It takes an array an returns that array in ascendent order
// Complexity: O(2^n)
function bubbleSort(array) {
  // We make a loop to repeat the ordering process until all the elements are ordered
  let unordered = true;
  while(unordered) {
    unordered = false;
    //We iterate through the whole array, comparing each element with the following
    // We have to stop at the second to last element so we don't have an out of bounds error
    for(let j = 0; j<array.length-1; j++) {
      // We compare each element with the following and if they are unordered we swap them
      if(array[j] > array[j+1]) {
        const aux = array[j+1];
        array[j+1] = array[j];
        array[j] = aux;
        //If we have made changes to the array we set unordered to true so we continue with the loop
        unordered = true
      }
    }
  }
  return array
}

const example = [2,54,12,3,5]
const orderedExample = bubbleSort(example);
console.log(orderedExample); // [2,3,5,12,54]